### ISSUE TITLE
functions reversed in zip output

### ISSUE THREAD

**Issue by gk-sysdig** [microsoft/pyright#9449]
In the below snippet, I get a pyright error indicating I'm missing parameters but it seems to be confusing the functions.

```python
import logging

logger = logging.getLogger(__name__)
logging.basicConfig(level=logging.DEBUG)

def func() -> int:
    logger.debug("in func...")
    return 3

def validator(old:int, new:int) -> bool:
    logger.debug(f"in validator. {old=}, {new=}")
    return old == new

funcs_to_test = (
    (
        func,
        validator,
    ),
    (
        func,
        validator,
    ),
)

queries, validators = zip(*funcs_to_test)
logger.debug(f'queries: {queries}')
logger.debug(f'validators: {validators}')

res = tuple(f() for f in queries)
print(res)

```

If I run this code, I get:
```
DEBUG:__main__:queries: (<function func at 0x10241c9a0>, <function func at 0x10241c9a0>)
DEBUG:__main__:validators: (<function validator at 0x1025876a0>, <function validator at 0x1025876a0>)
DEBUG:__main__:in func...
DEBUG:__main__:in func...
(3, 3)
```

Pyright error output:
```
/Users/user/test_pyright4.py
  /Users/user/test_pyright4.py:29:13 - error: Arguments missing for parameters "old", "new" (reportCallIssue)
1 error, 0 warnings, 0 informations
```

I am using pyright 1.1.388

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug.

The issue is with the way the `zip` class is defined in the typeshed stubs. Here's the overload that applies in your situation:

```python
class zip[T]:
        ...
        def __new__[T1, T2](cls, iter1: Iterable[T1], iter2: Iterable[T2], /, *, strict: bool = ...) -> zip[tuple[T1, T2]]: ...
```

This definition assumes homogeneous iterable types. You're passing heterogeneous iterables (tuples with two different callable types). The typeshed definition results in the heterogeneous iterable types being combined via a union.

There has been discussion in the typing forums of extending the type system to support this case better, but the discussions haven't yet coalesced into a formal proposal. 

In the meantime, you will need to work around this limitation. You could use a `# type: ignore` or `# pyright: ignore` comment, or you could use a type annotation like `funcs_to_test: tuple[Any, ...] = ...`.

**Comment by gk-sysdig** [comment]
Thanks for the analysis, explanation and workaround!
